src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManager.cs

```diff
@@ -0,0 +1,228 @@
+    /// amount of new work that can be fetched. The manager internally adjusts concurrency based on various
+    /// health heuristics.
+    /// </summary>
+    public class ConcurrencyManager
+    {
+        internal const int MinConsecutiveIncreaseLimit = 5;
+        internal const int MinConsecutiveDecreaseLimit = 3;
+
+        private readonly ILogger _logger;
+        private readonly IOptions<ConcurrencyOptions> _options;
+        private readonly IConcurrencyThrottleManager _concurrencyThrottleManager;
+
+        private ConcurrencyThrottleAggregateStatus _throttleStatus;
+
+        // for mock testing only
+        internal ConcurrencyManager()
+        {
+        }
+
+        public ConcurrencyManager(IOptions<ConcurrencyOptions> options, ILoggerFactory loggerFactory, IConcurrencyThrottleManager concurrencyThrottleManager)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+            if (concurrencyThrottleManager == null)
+            {
+                throw new ArgumentNullException(nameof(concurrencyThrottleManager));
+            }
+
+            _options = options;
+            _logger = loggerFactory.CreateLogger(LogCategories.Concurrency);
+            _concurrencyThrottleManager = concurrencyThrottleManager;
+
+            EffectiveCoresCount = Utility.GetEffectiveCoresCount();
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether dynamic concurrency is enabled.
+        /// </summary>
+        public bool Enabled => _options.Value.DynamicConcurrencyEnabled;
+
+        /// <summary>
+        /// Gets a value indicating whether any throttles are currently enabled.
+        /// </summary>
+        internal virtual bool ThrottleEnabled => _throttleStatus?.State == ThrottleState.Enabled;
+
+        internal int EffectiveCoresCount { get; set; }
+
+        internal ConcurrentDictionary<string, ConcurrencyStatus> ConcurrencyStatuses = new ConcurrentDictionary<string, ConcurrencyStatus>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Get the concurrency status for the specified function. 
+        /// </summary>
+        /// <param name="functionId">This should be the full ID, as returned by <see cref="FunctionDescriptor.ID"/>.</param>
+        /// <returns>The updated concurrency status.</returns>
+        /// <remarks>
+        /// Update shouldn't be called concurrently for the same function ID.
+        /// </remarks>
+        public ConcurrencyStatus GetStatus(string functionId)
+        {
+            if (string.IsNullOrEmpty(functionId))
+            {
+                throw new ArgumentNullException(nameof(functionId));
+            }
+
+            // because Update won't be called concurrently for the same function ID, we can make
```

Comment:
What is the "Update" that this comment is referring to?

Reply:
Referring to this method - GetStatus. Will update comment